File: enum-mybatis-validation-app/src/main/java/com/example/enumapp/web/validation/NotBlankFieldsInListValidator.java
package com.example.enumapp.web.validation;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;
import org.springframework.beans.BeanWrapperImpl;

import java.util.Collection;

public class NotBlankFieldsInListValidator implements ConstraintValidator<NotBlankFieldsInList, Collection<?>> {

    private String[] fields;

    @Override
    public void initialize(NotBlankFieldsInList annotation) {
        this.fields = annotation.fields();
    }

    @Override
    public boolean isValid(Collection<?> value, ConstraintValidatorContext context) {
        if (value == null) {
            context.disableDefaultConstraintViolation();
            context.buildConstraintViolationWithTemplate("list must not be null")
                    .addConstraintViolation();
            return false;
        }
        if (value.isEmpty()) {
            context.disableDefaultConstraintViolation();
            context.buildConstraintViolationWithTemplate("list must not be empty")
                    .addConstraintViolation();
            return false;
        }

        int index = 0;
        boolean valid = true;
        for (Object item : value) {
            if (item == null) {
                context.disableDefaultConstraintViolation();
                context.buildConstraintViolationWithTemplate("list[%d] must not be null".formatted(index))
                        .addConstraintViolation();
                valid = false;
                index++;
                continue;
            }
            BeanWrapperImpl wrapper = new BeanWrapperImpl(item);
            for (String field : fields) {
                Object fieldValue = wrapper.getPropertyValue(field);
                if (fieldValue == null || (fieldValue instanceof String s && s.isBlank())) {
                    context.disableDefaultConstraintViolation();
                    context.buildConstraintViolationWithTemplate(
                                    "list[%d].%s must not be null or blank".formatted(index, field)
                            )
                            .addConstraintViolation();
                    valid = false;
                }
            }
            index++;
        }
        return valid;
    }
}

File: enum-jpa-querydsl-kotlin-app/src/main/kotlin/com/example/enumapp/web/validation/NotBlankFieldsInListValidator.kt
package com.example.enumapp.web.validation

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import org.springframework.beans.BeanWrapperImpl

class NotBlankFieldsInListValidator : ConstraintValidator<NotBlankFieldsInList, Collection<*>?> {

    private lateinit var fields: Array<String>

    override fun initialize(annotation: NotBlankFieldsInList) {
        fields = annotation.fields
    }

    override fun isValid(value: Collection<*>?, context: ConstraintValidatorContext): Boolean {
        if (value == null) {
            context.disableDefaultConstraintViolation()
            context.buildConstraintViolationWithTemplate("list must not be null")
                .addConstraintViolation()
            return false
        }
        if (value.isEmpty()) {
            context.disableDefaultConstraintViolation()
            context.buildConstraintViolationWithTemplate("list must not be empty")
                .addConstraintViolation()
            return false
        }

        var index = 0
        var valid = true
        for (item in value) {
            if (item == null) {
                context.disableDefaultConstraintViolation()
                context.buildConstraintViolationWithTemplate("list[%d] must not be null".format(index))
                    .addConstraintViolation()
                valid = false
                index++
                continue
            }
            val wrapper = BeanWrapperImpl(item)
            for (field in fields) {
                val fieldValue = wrapper.getPropertyValue(field)
                if (fieldValue == null || (fieldValue is String && fieldValue.isBlank())) {
                    context.disableDefaultConstraintViolation()
                    context.buildConstraintViolationWithTemplate(
                        "list[%d].%s must not be null or blank".format(index, field),
                    )
                        .addConstraintViolation()
                    valid = false
                }
            }
            index++
        }
        return valid
    }
}
